autorest/azure/async_test.go

```diff
@@ -812,6 +811,112 @@ func TestDoPollForAsynchronous_ReturnsAnErrorForFailedOperations(t *testing.T) {
+	// Return error code and message if error present in last reponse
+	r1 := newAsynchronousResponse()
+	r1.Header.Del(http.CanonicalHeaderKey(headerAsyncOperation))
+	r2 := newProvisioningStatusResponse("busy")
+	r2.Header.Del(http.CanonicalHeaderKey(headerAsyncOperation))
+	r3 := newAsynchronousResponseWithError()
+	r3.Header.Del(http.CanonicalHeaderKey(headerAsyncOperation))
+
+	client := mocks.NewSender()
+	client.AppendResponse(r1)
+	client.AppendAndRepeatResponse(r2, 2)
+	client.AppendAndRepeatResponse(r3, 1)
+
+	r, err := autorest.SendWithSender(client, mocks.NewRequest(),
+		DoPollForAsynchronous(time.Millisecond))
+
+	if err.Error() != getErrorString("NetcfgInvalidSubnet", "Subnet is not valid in virtual network.") {
+		t.Errorf("azure: DoPollForAsynchronous failed to return an appropriate error message for a failed Operations for unknown error")
+	}
+
+	autorest.Respond(r,
+		autorest.ByClosing())
+}
+
+func TestDoPollForAsynchronous_ReturnsOperationResourceErrorForFailedOperations(t *testing.T) {
+	// Return Operation resource response with error code and message in last operation resource response
+	r1 := newAsynchronousResponse()
+	r2 := newOperationResourceResponse("busy")
+	r3 := newOperationResourceErrorResponse(operationFailed)
+
+	client := mocks.NewSender()
+	client.AppendResponse(r1)
+	client.AppendAndRepeatResponse(r2, 2)
+	client.AppendAndRepeatResponse(r3, 1)
+
+	r, err := autorest.SendWithSender(client, mocks.NewRequest(),
+		DoPollForAsynchronous(time.Millisecond))
+
+	if err.Error() != getErrorString("BadArgument", "The provided database 'foo' has an invalid username.") {
+		t.Errorf("azure: DoPollForAsynchronous failed to return an appropriate error message for a failed Operations")
+	}
+
+	autorest.Respond(r,
+		autorest.ByClosing())
+}
+
+func TestDoPollForAsynchronous_ReturnsErrorForFirstPutWithUnknownError(t *testing.T) {
+	// Return 400 bad response for unknown error in first put
+	r1 := newAsynchronousErrorResponseWithUnknownError()
+	client := mocks.NewSender()
+	client.AppendResponse(r1)
+
+	r, err := autorest.SendWithSender(client, mocks.NewRequest(),
+		DoPollForAsynchronous(time.Millisecond))
+
+	if err.Error() != getErrorString("Unknown", "None") {
+		t.Errorf("azure: DoPollForAsynchronous failed to return an appropriate error message for a failed Operations")
+	}
+
+	autorest.Respond(r,
+		autorest.ByClosing())
+}
+
+func TestDoPollForAsynchronous_ReturnsErrorForFirstPutRequest(t *testing.T) {
+	// Return 400 bad response with error code and message in first put
+	r1 := newAsynchronousResponseWithError()
+	client := mocks.NewSender()
+	client.AppendResponse(r1)
+
+	r, err := autorest.SendWithSender(client, mocks.NewRequest(),
+		DoPollForAsynchronous(time.Millisecond))
+
+	if err.Error() != getErrorString("NetcfgInvalidSubnet", "Subnet is not valid in virtual network.") {
+		t.Errorf("azure: DoPollForAsynchronous failed to return an appropriate error message for a failed Operations")
```

Comment:
if you fail with `... expected=%q got=%q`, when it fails the person who's trying to understand the failure will be happier :)